docs/rfcs/039-single-click-pg-upgrade.md

```diff
@@ -0,0 +1,199 @@
+## How pg_upgrade works in upstream postgres:
+
+0. USER: Initdb new version cluster and prepare it (install extension shared object files to the right path, adjust authentication and config)
+1. USER: stop both clusters
+2. USER run pg_upgade
+
+3. pg_upgrade: pg_dump of old cluster
+
+```
+start_postmaster(&old_cluster)
+
+get_db_and_rel_infos(&old_cluster);
+init_tablespaces();
+get_loadable_libraries();
+
+// some other checks
+
+generate_old_dump();
+
+stop_postmaster()
+
+```
+
+4. pg_upgrade: check and prepare new cluster
+
+```
+start_postmaster(&new_cluster);
+check_new_cluster();
+    get_db_and_rel_infos(&new_cluster);
+	check_new_cluster_is_empty();
+	check_loadable_libraries();
+
+set_locale_and_encoding();
+
+// run `vacuumdb --all --analyze` and  `vacuumdb --all --freeze`
+prepare_new_cluster();
+
+stop_postmaster();
+
+```
+
+5. pg_upgrade: handle SLRU
+
+```
+copy_xact_xlog_xid()
+    copy pg_xact files
+    set the next transaction id and epoch of the new cluster 
+
+    copy multixact files
+    Setting next multixact ID and offset for new cluster
+```
+
+6. pg_upgrade: pg_restore to new cluster
+
+```
+    // Note that postgres starts in a special --binary-upgrade mode.
+	start_postmaster(&new_cluster, true);
+
+	prepare_new_globals();
+
+    // Basically pg_restore, but postgres runs in --binary-upgrade mode
+    // and preserves OIDs and relfilenodes of imported objects
+	create_new_objects();
+
+	stop_postmaster();
+```
+
+7. pg_upgrade: transfer data files
+
+
+#### Storage compatibility:
+
+1. OID and relfilenode:
+pg_upgrade ensures that all assignments of `pg_class.oid` (and `relfilenode`) so toast oids are the same between old and new clusters. 
+It also controls assignments of `pg_tablespace.oid`, `pg_type.oid`, `pg_enum.oid` and `pg_authid.oid`.
+
+To do this, it starts postgres in (undocumented) --binary-upgrade mode.
+
+2. Page layout of relation and SLRU
+
+Postgres aims to maintain backward compatibility of page layout and change it lazily in the new version.
+In exceptional cases, it provides a function to run during pg_upgrade the page layout.
```

Comment:
```suggestion
In exceptional cases, it provides a function to run during pg_upgrade to rewrite pages to the new format.
```

We've never done that for relation files, but it has been done for some SLRUs in the past.  v17->v18 might involve some SLRU rewriting, thanks to the patches to move to 64-bit xids.